pyat/at/lattice/variable_elements.py

```diff
@@ -1,64 +1,85 @@
-import numpy
+from __future__ import annotations
+
+from datetime import datetime
+from enum import IntEnum
+
+import numpy as np
+
 from .elements import Element, _array
 from .utils import AtError
-from typing import Optional, Union
-from enum import IntEnum
-from datetime import datetime
 
 
 class ACMode(IntEnum):
-    """Class to define the excitation types"""
+    """Class to define the excitation types."""
+
     SINE = 0
     WHITENOISE = 1
     ARBITRARY = 2
 
 
 class VariableMultipole(Element):
-    """Class to generate an AT variable thin multipole element
-    """
+    """Class to generate an AT variable thin multipole element."""
+
     _BUILD_ATTRIBUTES = Element._BUILD_ATTRIBUTES
-    _conversions = dict(Element._conversions,
-                        Mode=int,
-                        AmplitudeA=_array, AmplitudeB=_array,
-                        FrequencyA=float, FrequencyB=float,
-                        PhaseA=float, PhaseB=float,
-                        Seed=int, NSamplesA=int, NSamplesB=int,
-                        FuncA=_array, FuncB=_array, Ramps=_array,
-                        Periodic=bool)
-
-    def __init__(self, family_name, AmplitudeA=None, AmplitudeB=None,
-                 mode=ACMode.SINE, **kwargs):
+    _conversions = dict(
+        Element._conversions,
+        Mode=int,
+        AmplitudeA=_array,
+        AmplitudeB=_array,
+        FrequencyA=float,
+        FrequencyB=float,
+        PhaseA=float,
+        PhaseB=float,
+        Seed=int,
+        NSamplesA=int,
+        NSamplesB=int,
+        FuncA=_array,
+        FuncB=_array,
+        Ramps=_array,
+        Periodic=bool,
+    )
+
+    def __init__(self, family_name: str, **kwargs: dict[str, any]):
```

Comment:
Dear @swhite2401 , if I remove the annotation I get the following flake8 warning. This is the reason why I included, should I ignore it ?  I may be using flake8 with a wrong configuration or I may use another tool ?
```
/variable_elements.py:42:44: ANN003 Missing type annotation for **kwargs
```

Reply:
> Could you please restore the defaults for AmplitudeA/B and the mode? I think it is better if they appear in the API doc, in fact I would be tempted to add more of these keyword arguments rather than removing them

Dear @swhite2401 , the reason why I remove it is because of an incompatibility between AT matlab and this implementation of the element in pyat.
If I save a lattice in matlab and try to read it in pyat I get the error here below, I'll try to explain the reason in  detail, but, in few words: **one cannot have args and kwargs with the same name**

When reading an element from matlab it is initially put in a dictionary `elem_dict`, and then the `_BUILD_ATTRIBUTES` are stripped to pass them to the class instantiation as `elem_args`. See the lines below
https://github.com/atcollab/at/blob/268994f8214cb758714eae5a3f55e95e5e915ad4/pyat/at/load/utils.py#L368-L369

I could not put the AmplutudeA, AmplitudeB and Mode in the element BUILD_ATTRIBUTES because they are not garanteed to exist in matlab, see the following line
https://github.com/atcollab/at/blob/268994f8214cb758714eae5a3f55e95e5e915ad4/atmat/lattice/element_creation/atvariablemultipole.m#L1
therefore, the call to cls with args in necessarily incomplete and the AmplitudeA, AmplitudeB and Mode are passed in a dictionary together with args that have the same name. Python does not handle this mix well, it removes the elements from the kwargs and leaves unset the args.

There are ways to workaround this situation when args and kwargs have the same name, but they require changing the way the class is instantiated in utils.py. I decided not to do it because that will change how all elements are instantiated.

A second possibility was to rename the element args in pyat, for example amplitudeA instead of AmplitudeA i.e. lower case in the first letter to make one small change. However, this would break compatibility.

I decided to keep the element without any requirement as in AT matlab, which forced me to do checks on the input parameters before assigning them inside `__init__`. This wouldn't break compatibility with previous versions of pyat, and AT matlab files could be read because the element also has no requirements except for the FamName.

If you prefer, I could do as you asked. Keep the args and break backwards compatibility. Or force the user to provide AmplitudeA, AmplitudeB and Mode in matlab, which will make possible to include them in BUILD_ATTRIBUTES that will be removed in the input dictionary and set before the class is instantiated.

Or maybe there is another solution ?

```python
>>> ring = at.load_mat('test1.mat')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/oblanco/Documents/public/progs/pyenv/pyenvat/lib/python3.11/site-packages/at/load/matfile.py", line 202, in load_mat
    return Lattice(
           ^^^^^^^^
  File "/home/oblanco/Documents/public/progs/pyenv/pyenvat/lib/python3.11/site-packages/at/lattice/lattice_object.py", line 191, in __init__
    super(Lattice, self).__init__(elems)
  File "/home/oblanco/Documents/public/progs/pyenv/pyenvat/lib/python3.11/site-packages/at/lattice/lattice_object.py", line 1478, in params_filter
    for elem in elem_filter(params, *args):
  File "/home/oblanco/Documents/public/progs/pyenv/pyenvat/lib/python3.11/site-packages/at/load/matfile.py", line 140, in ringparam_filter
    for elem in elem_iterator(params, *args):
  File "/home/oblanco/Documents/public/progs/pyenv/pyenvat/lib/python3.11/site-packages/at/load/matfile.py", line 105, in _matfile_generator
    yield Element.from_dict(kwargs, index=index, check=check, quiet=quiet)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/oblanco/Documents/public/progs/pyenv/pyenvat/lib/python3.11/site-packages/at/load/utils.py", line 365, in element_from_dict
    element = cls(*(arg for arg in elem_args if arg is not None), **elem_dict)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/oblanco/Documents/public/progs/pyenv/pyenvat/lib/python3.11/site-packages/at/lattice/variable_elements.py", line 111, in __init__
    raise AtError("Please provide at least one amplitude for A or B")
at.lattice.utils.AtError: Please provide at least one amplitude for A or B
```

Reply:
Dear @swhite2401 , I asked @ZeusMarti  and he was not able to use it effectively because he intends to have a custom function where the time of the particle is also considered, i.e. not only a kick every turn for all particles.
I have started modifying the C code today to make some test, is this feature something you would accept for this element ? Or should I keep the custom function option as it is ?

o

Reply:
> I have started modifying the C code today to make some test, is this feature something you would accept for this element ?

Sure! In fact I developed this for @ZeusMarti initially, so any improvement is perfectly fine.
You can already do a kick every n turns by using the table by the way, just fill zeros where you do not want any kick.

Do you want help for this?